File: src/courses/data/catalog/02-dsa/advanced-dsa/implementations/java/trie.java
/**
 * Java Implementation: Trie (Prefix Tree)
 * Invariant: Prefix search O(Length of Word)
 */
package seed.dsa.core;

import java.util.*;

public class TriePrefixTree<T> {
    private final List<T> items = new ArrayList<>();

    // TODO: Implement your solution here
    public void add(T value) {
        items.add(value);
    }

    public int size() {
        return items.size();
    }
}

File: src/courses/data/catalog/02-dsa/advanced-dsa/implementations/cpp/trie.cpp
/**
 * C++ Implementation: Trie (Prefix Tree)
 * Invariant: Prefix search O(Length of Word)
 */
#include <iostream>
#include <vector>
#include <stdexcept>

template <typename T>
class TriePrefixTree {
private:
    std::vector<T> elements;
public:
    // TODO: Implement primary invariant operations
    void push(const T& val) {
        elements.push_back(val);
    }
    size_t size() const { return elements.size(); }
};
